### PR TITLE
Add configurable turn completion message

### LINKED ISSUE #66: Feature: configurable turn completion indicator for streamed agent replies
## Problem

A single agent turn can send several WeChat messages before it actually finishes. GPT models often emit multiple message segments around thoughts and tool calls. Long-running tools can then stay silent for a long time.

From WeChat, there is no reliable way to tell whether the latest message is the final answer or whether the same turn is still running. The bridge already knows the exact boundary because `connection.prompt()` resolves with a `stopReason`, and logs entries such as:

```text
Agent done (end_turn), reply 85 chars
```

That signal is currently visible only in the bridge log, not in WeChat.

## Goal

Provide an optional, clear, bridge-generated completion indicator when an ACP prompt turn ends.

The indicator must reflect the real ACP turn boundary. It should not depend on the model remembering to print a closing phrase.

## Recommended behavior

Add an optional config value such as:

```json
{
  "session": {
    "turnEndMessage": "✅ Turn complete"
  }
}
```

Also expose a CLI override for startup use:

```text
--turn-end-message <text>
```

An omitted or empty value disables the indicator. Starting as opt-in preserves current behavior and avoids adding noise for existing users.

When `connection.prompt()` resolves and `client.flush()` has drained all queued ACP notifications, deliver the configured text exactly once using this hybrid placement rule:

1. If `flush()` returns final reply text, append the indicator as the final paragraph of that same send.
2. If earlier message segments were already streamed and the final flush is empty, send the indicator as a standalone final message.
3. If the turn produced no text and the bridge sends `emptyTurnNotice(...)`, append the indicator to that notice.
4. Keep existing stop reason text such as `[cancelled]` and `[agent refused to continue]`, then place the completion indicator last.

This gives the user a stable invariant: once the configured marker appears, no more output belongs to that turn.

## Why hybrid placement

Always appending to the agent's final answer is not enough. `src/acp/client.ts` can flush message segments at thought or tool-call boundaries, so by the time `prompt()` resolves, the last visible agent message may already have been sent and `flush()` may return an empty string.

Always sending a separate message would work, but creates extra chat noise and consumes one more outbound message for every turn. The hybrid rule adds a separate message only when it is needed.

## Why this belongs in the bridge

Prompting the model to emit a closing phrase is not reliable. The model can stop because of cancellation, refusal, token limits, tool limits, an agent implementation detail, or a missed instruction. It may also print the phrase before asynchronous delivery has fully drained.

The bridge has the authoritative lifecycle signal at `await connection.prompt(...)` and can place the marker after `await client.flush()`. The existing per-user send queue in `src/bridge.ts` can preserve ordering.

### Keep the typing indicator active until completion

This could improve progress feedback, but it is not a durable completion signal. Typing indicators can expire, disappear after intermediate sends, or be missed when the user opens WeChat later. It could be a separate improvement, not a replacement for the marker.

### Always send a standalone completion message

This is simple and visually clear, but adds one message to every turn and makes WeChat's per-inbound outbound cap more likely. See issue #45.

### Modify the last message after completion

That would provide the cleanest UI if WeChat supported reliable message edits for this bot API. The current send path only creates messages, so this does not appear available.

## Delivery limits

The indicator is subject to the same WeChat delivery failures and per-inbound outbound quota as other messages. Appending it to a non-empty final flush reduces the extra quota cost. If the quota has already been exhausted before the turn ends, this feature alone cannot guarantee delivery. Issue #45 tracks recovery of dropped reply segments using a fresh inbound `context_token`.

## Implementation outline

- Add `session.turnEndMessage?: string` to `WeChatAcpConfig` and its default config.
- Parse `--turn-end-message <text>` with CLI precedence over the config file.
- Pass the resolved value through `WeChatAcpBridge` into `SessionManagerOpts`.
- In `SessionManager.processQueue()`, add the marker only after `prompt()` resolves and `client.flush()` completes.
- Reuse `onReply` so existing per-user serialization and message spacing preserve order.
- Do not ask the ACP agent or model to generate the marker.
- Document the config and CLI option.

## Acceptance criteria

- With no config or CLI option, behavior is unchanged.
- A configured marker is delivered exactly once for each resolved ACP prompt turn.
- The marker is always ordered after all message, thought, image, audio, file, and resource output from that turn.
- A turn whose final text remains buffered gets the marker appended to that final text.
- A turn whose text was fully streamed before `prompt()` resolves gets a standalone marker.
- Cancelled, refused, and empty-text turns keep their current notices and end with the configured marker.
- Queued user prompts each receive their own marker, with no marker leaking into the next turn.
- Tests cover final buffered text, fully streamed text, empty output, cancellation, refusal, and send failure.

### PR DESCRIPTION
**Pull request by Eskibear**
## Summary

- add optional `session.turnEndMessage` configuration
- add a `--turn-end-message <text>` CLI override
- send the configured text as a standalone message after the ACP prompt and queued turn output finish
- keep existing behavior unchanged when no completion message is configured

Fixes #66.

## Tests

- `npm run build`
- `node --import tsx/esm --test tests/session.test.ts`
- full test runner: 116 passed; the existing artifact symlink test could not create a symlink because this Windows environment lacks permission

### CODE REVIEW

File: tests/session.test.ts [hunk 1]
```diff
@@ -120,3 +120,132 @@ test("SessionManager.stop waits for every MCP lease before reporting failures",
...
+    },
+    configOptions: [],
+    queue: [{ prompt: [], contextToken: "turn-context" }],
+    processing: true,
+    lastActivity: Date.now(),
+    createdAt: Date.now(),
+  };
+}
+
+test("configured turn end message is sent standalone after final buffered text", async () => {
```
[the remark this answers is not in the document]
Comment (Eskibear) [reply tests/session.test.ts:170]: Intentional. We agreed in issue #66 to start with an always-standalone completion message. This keeps the marker visually distinct and the implementation simple. We can revisit hybrid placement if quota or chat noise becomes a real problem.

File: README.md [hunk 2]
```diff
@@ -101,6 +101,7 @@ Options:
 - `--instance <name>`: run as a named, isolated instance. See "Running multiple instances" below.
 - `--idle-timeout <minutes>`: session idle timeout, default `1440` (use `0` for unlimited)
 - `--max-sessions <count>`: maximum concurrent user sessions, default `10`
+- `--turn-end-message <text>`: send a standalone message after each completed agent turn (default: disabled)
```
[the remark this answers is not in the document]
Comment (Eskibear) [reply README.md:104]: Intentional. We agreed in issue #66 to start with an always-standalone completion message. This keeps the marker visually distinct and the implementation simple. We can revisit hybrid placement if quota or chat noise becomes a real problem.

File: README.md [hunk 3]
```diff
@@ -161,11 +162,20 @@ Example:
...
 }
 ```
 
+`session.turnEndMessage` is optional. When set to a non-empty string, the
+bridge sends it as a standalone WeChat message after the ACP prompt resolves
+and all output from that turn has been delivered. This makes the real turn
+boundary visible even when the agent streamed several earlier messages or was
+silent during a long-running tool call. The bridge generates this signal, so it
+does not depend on the model following a prompt instruction. The
+`--turn-end-message` CLI option overrides the config file value.
```
[the remark this answers is not in the document]
Comment (Eskibear) [reply README.md:177]: Intentional. We agreed in issue #66 to start with an always-standalone completion message. This keeps the marker visually distinct and the implementation simple. We can revisit hybrid placement if quota or chat noise becomes a real problem.

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 ## 0.10.0
 
+- Add an opt-in standalone turn completion message so WeChat users can tell when a streamed or long-running ACP prompt has actually ended. Configure `session.turnEndMessage` or pass `--turn-end-message <text>`; the bridge sends the text after `prompt()` resolves and the turn's queued output has drained. Fixes #66.
```
[the remark this answers is not in the document]
Comment (Eskibear) [reply CHANGELOG.md:5]: Intentional. We agreed in issue #66 to start with an always-standalone completion message. This keeps the marker visually distinct and the implementation simple. We can revisit hybrid placement if quota or chat noise becomes a real problem.

File: src/acp/session.ts [hunk 5]
```diff
@@ -485,6 +486,14 @@ export class SessionManager {
...
             );
           }
+
+          if (this.opts.turnEndMessage?.trim()) {
+            await this.opts.onReply(
+              session.userId,
+              pending.contextToken,
+              this.opts.turnEndMessage,
+            );
+          }
```
[the remark this answers is not in the document]
Comment (Eskibear) [reply src/acp/session.ts:503]: Intentional. We agreed in issue #66 to start with an always-standalone completion message. This keeps the marker visually distinct and the implementation simple. We can revisit hybrid placement if quota or chat noise becomes a real problem.

File: src/config.ts [hunk 6]
```diff
@@ -151,6 +151,11 @@ export interface WeChatAcpConfig {
   session: {
     idleTimeoutMs: number;
     maxConcurrentUsers: number;
+    /**
+     * Optional standalone message sent after an ACP prompt turn completes.
+     * Omit or set to an empty string to disable the completion indicator.
+     */
```
[the remark this answers is not in the document]
Comment (Eskibear) [reply src/config.ts:157]: Intentional. We agreed in issue #66 to start with an always-standalone completion message. This keeps the marker visually distinct and the implementation simple. We can revisit hybrid placement if quota or chat noise becomes a real problem.

File: tests/session.test.ts [hunk 7]
```diff
@@ -120,3 +120,132 @@ test("SessionManager.stop waits for every MCP lease before reporting failures",
...
+    producedMessage: true,
+    events: [],
+  });
+
+  await (
+    manager as unknown as { processQueue(session: UserSession): Promise<void> }
+  ).processQueue(session);
+
+  assert.deepEqual(replies, ["Done"]);
+});
```
[the remark this answers is not in the document]
Comment (Eskibear) [reply tests/session.test.ts:232]: Added coverage for cancelled, refusal, and empty-output paths in `65d164c`. I did not add a synthetic `onReply` failure case because completion uses the existing delivery path, whose retry and failure behavior is tested separately.

File: CHANGELOG.md (hunk 4 above)
Comment (formulahendry) [inline CHANGELOG.md:5]: Minor: `v0.10.0` was already released on July 25, 2026, so adding this unreleased feature under that heading makes the published release history inaccurate. Could this entry go under a new `Unreleased` section (and be renamed to the eventual release version, likely `0.11.0`, during release) instead?

File: src/acp/session.ts [hunk 8]
```diff
@@ -485,6 +486,14 @@ export class SessionManager {
               emptyTurnNotice(result.stopReason),
             );
           }
+
+          if (this.opts.turnEndMessage?.trim()) {
+            await this.opts.onReply(
```
Comment (formulahendry) [inline src/acp/session.ts]: Minor: This `onReply()` runs inside the broad Agent prompt `try/catch`. If the completion-marker callback rejects after `prompt.completed` has already been emitted with `success: true`, control falls into the prompt-error path, emits a conflicting failure event, attempts to send `⚠️ Agent error`, and rejects the turn completion even though the agent response itself succeeded. Could we catch marker delivery separately and classify/log it as a reply-delivery failure instead? A regression test where only the marker send rejects would pin this behavior down.

**Review by formulahendry — COMMENTED** [review]
See inline comments.

File: CHANGELOG.md (hunk 4 above)
Comment (Eskibear) [reply CHANGELOG.md:5]: Good catch. Moved the entry to a new `Unreleased` section in `6ebda86`.

File: src/acp/session.ts (hunk 8 above)
Comment (Eskibear) [reply src/acp/session.ts]: Fixed in `6ebda86`. Marker delivery now has its own error boundary and `reply.turn_end` telemetry, so it cannot emit a false prompt failure or reject a completed turn. Added a regression test for this path.

**Review by formulahendry — APPROVED** [review]
The previously reported issues are resolved. The remaining malformed-config edge case is non-blocking; normal string values, including Unicode and emoji, are handled correctly.
